### PR TITLE
feat: middleware implementation

### PR DESCRIPTION
**Pull request by Industrial**

### CODE REVIEW

**Comment by deckchairlabs** [comment]
Looks like our server processes are never closing when running puppeteer tests.

**Comment by deckchairlabs** [comment]
Fixed the e2e issues, so tests should pass now

**Comment by deckchairlabs** [comment]
Should shortCircuit just be a response? Like if I wanted to just short circuit the middleware chain, I could just pass a new response to next?

**Comment by deckchairlabs** [comment]
Is there a way to modify a response thats been handled by an internal middleware?

**Comment by dburles** [comment]
We should just expose `requestHandler` as middleware itself

File: src/server/middleware/transpileSource.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  ValidExtensions,
+} from "../../resolver.ts";
+import { resolveConfig, resolveImportMap } from "../../config.ts";
+
+export default async function transpileSource(): Promise<Middleware> {
+  const cwd = Deno.cwd();
+  const config = await resolveConfig(cwd);
+  const importMap = await resolveImportMap(cwd, config);
+  const memory = new LRU<string>(500);
+  const rawAssets = await assets(sourceDirectory);
```
Comment (dburles) [inline src/server/middleware/transpileSource.ts:19]: `staticAsset` also runs this function, it should run once for best perf.

File: src/server/middleware/transpileSource.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import { createURL } from "../request.ts";
+import { isDev, sourceDirectory } from "../../env.ts";
+import {
+  replaceFileExt,
+  resolveFileUrl,
+  ValidExtensions,
+} from "../../resolver.ts";
+import { resolveConfig, resolveImportMap } from "../../config.ts";
+
+export default async function transpileSource(): Promise<Middleware> {
```
Comment (dburles) [inline src/server/middleware/transpileSource.ts:14]: Rename these functions that return middleware to: `createTranspileSourceMiddleware` and return named functions like `transpileSourceMiddleware`.

File: workspace/server.js [hunk 3]
```diff
@@ -1,3 +1,16 @@
...
+
+server.use(async (context, next) => {
+  console.log(`<-- ${context.request.method} ${context.request.url}`);
+  await next();
+  console.log(
+    `--> ${context.request.method} ${context.request.url} ${context.response.status}`,
+  );
+});
+
+server.use(requestHandler());
```
Comment (dburles) [inline workspace/server.js:14]: Ideally this should be the middleware function at this stage and not a factory
```suggestion
server.use(requestHandler);
```

**Review by dburles — CHANGES_REQUESTED** [review]

File: workspace/server.js (hunk 3 above)
Comment (dburles) [reply workspace/server.js:14]: Otherwise (if necessary) `createRequestHandler` will make it clearer.

Comment (Industrial) [reply workspace/server.js:14]: If passing middleware functions directly without factories then there can be no scope outside the request scope. This means that some resources will get re-created for every request. So I think factories are necessary.

Comment (dburles) [reply workspace/server.js]: You can just initialise those outside of function scope at the root instead, the only time that can be an issue though is with circular deps.

File: src/server/middleware/createRequestHandlerMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  return compose(
+    createTranspileSourceMiddleware(
+      rawAssets,
+      importMap,
+    ),
+    createStaticAssetMiddleware(rawAssets),
+    createVendorMapMiddleware(vendorAssets),
+    createRenderPageMiddleware(importMap),
+  );
+}
```
Comment (dburles) [inline src/server/middleware/createRequestHandlerMiddleware.ts:38]: This is what I mean, you can just do this:

```suggestion

const [
  rawAssets,
  vendorAssets,
  importMap,
] = await Promise.all([
  assets(sourceDirectory),
  assets(`.ultra/${vendorDirectory}`),
  (async () => {
    const cwd = Deno.cwd();
    const config = await resolveConfig(cwd);
    const importMap = await resolveImportMap(cwd, config);
    return importMap;
  })(),
]);

const requestHandlerMiddleware = compose(
  createTranspileSourceMiddleware(
    rawAssets,
    importMap,
  ),
  createStaticAssetMiddleware(rawAssets),
  createVendorMapMiddleware(vendorAssets),
  createRenderPageMiddleware(importMap),
);

export default requestHandlerMiddleware;
```

**Pull request closed without merging**